linchpin/api/__init__.py

```diff
@@ -1,28 +1,77 @@
 #!/usr/bin/env python
 
 import os
-
 from linchpin.api.invoke_playbooks import invoke_linchpin
 from linchpin.api.utils import yaml2json
+from linchpin.hooks import LinchpinHooks
+from linchpin.hooks.state import State
 
 
 class LinchpinError(Exception):
 
     def __init__(self,*args,**kwargs):
         Exception.__init__(self,*args,**kwargs)
 
-class LinchpinAPI:
+
+class LinchpinAPI(object):
 
     def __init__(self, ctx):
 
         self.ctx = ctx
         base_path = '/'.join(os.path.dirname(__file__).split("/")[0:-2])
         self.lp_path = '{0}/{1}'.format(base_path, self.ctx.cfgs['lp']['pkg'])
+        self._state = None
+        self._state_observers = []
+        self.hooks = LinchpinHooks(self)
+        self.current_target_data = {}
+
+    @property
+    def state(self):
+        """ getter function for state property of the API object. """
+        return self._state
+
+    @state.setter
+    def state(self, state):
+        # call run_hooks after state is being set
+        print("State change initiated")
```

Comment:
Remove print statements. If wanted, use ctx.log, ctx.log_state (which outputs to stderr), ctx.log_info, or ctx.log_debug. If that's not enough, more can be created in the context.py.